File: fsharp/palindrome-products/PalindromeProducts.fs
module PalindromeProducts

let private isPalindrome num = 
    let rec reverse (num,res) =
        match num with
        | 0 -> res
        | _ -> reverse (num/10,res * 10 + num % 10)
    reverse (num, 0) |> (=) num

let private palindromes minFactor maxFactor = 
    Seq.collect (fun a -> [minFactor..a] |> Seq.map (fun b -> a*b)) [minFactor..maxFactor]
    |> Seq.filter isPalindrome

let private factorise minFactor maxFactor num =
    let factors = 
        [minFactor..maxFactor] |> Seq.map (fun a -> (a, num / a, num % a))
        |> Seq.filter (fun (_,b,m) -> m = 0 && b >= minFactor && b <= maxFactor)
        |> Seq.map (fun (a,b,_) -> if a < b then (a,b) else (b,a)) |> Seq.distinct |> Seq.toList
    (num, factors)

let largest minFactor maxFactor = 
    let set = palindromes minFactor maxFactor
    if Seq.isEmpty set then None else set |> Seq.max |> factorise minFactor maxFactor |> Some

let smallest minFactor maxFactor = 
    let set = palindromes minFactor maxFactor
    if Seq.isEmpty set then None else set |> Seq.min |> factorise minFactor maxFactor |> Some

File: csharp/palindrome-products/PalindromeProducts.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class PalindromeProducts
{
    public static (int, IEnumerable<(int,int)>) Largest(int minFactor, int maxFactor)
        => Result(minFactor, maxFactor, false);

    public static (int, IEnumerable<(int,int)>) Smallest(int minFactor, int maxFactor)
        => Result(minFactor, maxFactor, true);

    public static (int, IEnumerable<(int,int)>) Result(int minFactor, int maxFactor, bool smallest)
    {
        if(minFactor >= maxFactor)
            throw new ArgumentException();

        var result = Palindromes(minFactor, maxFactor)
            .GroupBy(o => o.product)
            .OrderBy(o => smallest ? o.Key : -o.Key)
            .First();

        return (result.Key, result.Select(o => (o.f1, o.f2)));
    }

    private static IEnumerable<(int f1, int f2, int product)> Palindromes(int minFactor, int maxFactor)
    {
        IEnumerable<int> range() => Enumerable.Range(minFactor, maxFactor - minFactor + 1);
        
        var results =
            range().SelectMany(f1 => range().Select(f2 => 
            {
                var first = f1 < f2 ? f1 : f2;
                var second = f2 > f1 ? f2 : f1;
                return (f1: first, f2: second, product: f1 * f2);
            }))
            .Where(o => IsPalindrome(o.product))
            .Distinct();

        if(!results.Any())
            throw new ArgumentException();

        return results;
    }

    private static bool IsPalindrome(int number)
    {
        var reverse = 0;
        var original = number;

        while(number > 0)
        {
            reverse = reverse * 10 + number % 10;
            number /= 10;
        }

        return reverse == original;
    }
}
